dartagnan/src/main/java/com/dat3m/dartagnan/program/analysis/alias/InclusionBasedPointerAnalysis.java

```diff
@@ -0,0 +1,897 @@
+
+    (*) We also generate special phi nodes that can improve precision on code that is not in SSA form.
+    (**) In the implementation we make an unsound assumption that x is non-negative.
+    This disallows negative (dynamic) indexing into arrays/pointers but gives additional precision in other cases.
+    Note that "o" can be negative which is sufficient to support "containerof" functionality.
+*/
+public class InclusionBasedPointerAnalysis implements AliasAnalysis {
+
+    private static final Logger logger = LogManager.getLogger(InclusionBasedPointerAnalysis.class);
+
+    private static final List<Integer> TOP = List.of(1);
+
+    // Debug: count inclusion tests grouped by complexity.
+    private final Map<Integer, int[]> totalAlignmentSizes = logger.isDebugEnabled() ? new HashMap<>() : null;
+    // Debug: count created variables.
+    private int totalVariables = 0;
+    // Debug: count variable substitutions.
+    private int totalReplacements = 0;
+    // Debug: count new edges already covered by existing ones.
+    private int failedAddInto = 0;
+    // Debug: count times a piece of new information was added to the graph.
+    private int succeededAddInto = 0;
+
+    // If enabled, the algorithm describes its internal graph to be written into a file.
+    private Graphviz graphviz;
+
+    // When a variable gains an includes-edge, it is added to this queue for later processing.
+    private final LinkedHashMap<Variable, List<Offset<Variable>>> queue = new LinkedHashMap<>();
+
+    // Maps memory events to variables representing their pointer set.
+    private final Map<MemoryCoreEvent, Offset<Variable>> addressVariables = new HashMap<>();
+
+    // Maps memory objects to variables representing their base address.
+    // These Variables should always have empty includes-sets.
+    private final Map<MemoryObject, Variable> objectVariables = new HashMap<>();
+
+    // Maps dependent sets of a register to variables representing their phi node's value set.
+    // Singletons are mapped to their output set.
+    // Offsets are for Local output sets; non-singletons and singletons of loads should have zero offset.
+    private final Map<List<RegWriter>, Offset<Variable>> registerVariables = new HashMap<>();
+
+    // This analysis depends on the results of a Dependency analysis, mapping used registers to a list of possible writers.
+    private final Dependency dependency;
+
+    // For providing helpful error messages, this analysis prints call-stack and loop information for events.
+    private final Supplier<SyntacticContextAnalysis> synContext;
+
+    private static final class Variable {
+        // Any value contained in these Variables is also contained (+ offset).
+        private final List<Offset<Variable>> includes = new ArrayList<>();
+        // There is some load event using this (+ offset) as pointer set and the other variable as output set.
+        private final List<Offset<Variable>> loads = new ArrayList<>();
```

Comment:
In the `includes`, a `Offset<Variable>` describes an incoming edge but in `loads` it is an outgoing edge.
This is confusing. Together with the fact that `Offset<Variable>` is sometimes also treated like a variable, I think this very hard to understand.